// @itr-app/web/src/pages/filing/itr1/ITR1Filing.jsx
import React, { useState, useEffect } from "react";
import {
  Steps, Card, Button, Form, Input, InputNumber,
  Select, DatePicker, Row, Col, Typography,
  Alert, Divider, Result, Tag, Space, Drawer,
  Tooltip, Spin, Grid, Upload, message,
  theme as antdTheme,
} from "antd";
import {
  UserOutlined, BankOutlined, FileTextOutlined,
  CheckCircleOutlined, ArrowLeftOutlined, ArrowRightOutlined,
  FilePdfOutlined, PaperClipOutlined, DownloadOutlined,
  SafetyCertificateOutlined, GlobalOutlined, InboxOutlined,
  ThunderboltOutlined, UploadOutlined,
} from "@ant-design/icons";
import { useAuthStore } from "../../../store/index.js";
import { useFilingStore } from "../../../store/index.js";
import { compareRegimes } from "../../../services/tax.service.js";
import { DEDUCTION_LIMITS, METRO_CITIES, isValidAadhaarChecksum } from "@itr-app/shared-types";
import { saveDraft, submitITR1, downloadFilingXML } from "../../../services/filing.service.js";
import { getMyDocuments, uploadDocument } from "../../../services/document.service.js";
import { useNavigate } from "react-router-dom";
import PageHeader from "../../../components/PageHeader.jsx";

const { Dragger } = Upload;

const { Title, Text } = Typography;
const { Option }      = Select;
const { useBreakpoint } = Grid;

const fmt = (n) =>
  new Intl.NumberFormat("en-IN", { style: "currency", currency: "INR", maximumFractionDigits: 0 })
    .format(n || 0);

const STEPS = [
  { title: "Personal",    icon: <UserOutlined />,        description: "Your details"     },
  { title: "Income",      icon: <BankOutlined />,         description: "Salary & others"  },
  { title: "Deductions",  icon: <FileTextOutlined />,     description: "80C, 80D & more"  },
  { title: "Tax Summary", icon: <CheckCircleOutlined />,  description: "Review & file"    },
];

export default function ITR1Filing() {
  const { token }                        = antdTheme.useToken();
  const { user }                         = useAuthStore();
  const { updateFiling, filingData }     = useFilingStore();
  const [current, setCurrent]            = useState(0);
  const [form]                           = Form.useForm();
  const [taxResult, setTaxResult]        = useState(null);
  const [loading, setLoading]            = useState(false);
  const [error, setError]                = useState(null);
  const navigate                         = useNavigate();
  const [submitLoading, setSubmitLoading] = useState(false);
  const [submitted, setSubmitted]         = useState(null);
  const screens                          = useBreakpoint();

  // ── Form 16 drawer state ─────────────────────────────────────
  const [drawerOpen, setDrawerOpen]   = useState(false);
  const [form16Doc, setForm16Doc]     = useState(null);
  const [docsLoading, setDocsLoading] = useState(false);
  const [uploading, setUploading]     = useState(false);

  // ── XML download state ────────────────────────────────────────
  const [xmlLoading, setXmlLoading]   = useState(false);
  const [xmlError,   setXmlError]     = useState(null);

  useEffect(() => {
    setDocsLoading(true);
    getMyDocuments()
      .then((res) => {
        const docs   = res.data || [];
        const latest = docs.find(
          (d) => d.type === "form16" && d.parseStatus === "success"
        );
        setForm16Doc(latest || null);
      })
      .catch(() => {})
      .finally(() => setDocsLoading(false));
  }, []);

  // Sync user identity fields into the form whenever the user object loads.
  // initialValue only fires on first mount — if user was null at that point
  // (page refresh before /auth/me resolves), pan and fullName stay empty.
  useEffect(() => {
    if (user?.fullName) form.setFieldValue("fullName", user.fullName);
    if (user?.pan)      form.setFieldValue("pan",      user.pan);
  }, [user]);

  // ── Form 16 upload + pre-fill ──────────────────────────────────
  const handleForm16Upload = async ({ file }) => {
    setUploading(true);
    try {
      const formData = new FormData();
      formData.append("document", file);
      formData.append("type", "form16");
      formData.append("financialYear", "2025-26");
      const res = await uploadDocument(formData);
      setForm16Doc(res.data);
      message.success("Form 16 uploaded and parsed!");
    } catch (err) {
      message.error(err.message || "Upload failed");
    } finally {
      setUploading(false);
    }
    return false;
  };

  const handleUseForm16Values = () => {
    const p = form16Doc?.parsedData;
    if (!p) return;
    // Single form instance with preserve: true (AntD default) keeps values for
    // fields belonging to steps that aren't currently mounted, so this is safe
    // to call regardless of which step the user is on.
    form.setFieldsValue({
      employerName:     p.employerName     || undefined,
      employerTAN:      p.employerTAN      || undefined,
      basicSalary:       p.basicSalary       || undefined,
      hra_received:      p.hraReceived       || undefined,
      specialAllowance:  p.specialAllowance  || undefined,
      bonus:             p.bonus             || undefined,
      professionalTax:   p.professionalTax   || undefined,
      tdsDeducted:       p.tdsDeducted       || undefined,
    });
    message.success("Form 16 values applied — review them in the Personal and Income steps");
    setDrawerOpen(false);
  };

  // Fields that belong to each step — used to scope validateFields() so that
  // preserved (unmounted) fields from other steps are not re-validated.
  const STEP_FIELDS = [
    ["fullName", "pan", "dateOfBirth", "gender", "residentialStatus",
     "fatherName", "aadhaar", "mobile", "addressLine1", "pinCode",
     "city", "employerName", "employerTAN", "bankAccountNo", "ifscCode"],
    ["basicSalary", "hra_received", "specialAllowance", "bonus", "professionalTax", "tdsDeducted", "interestIncome", "otherIncome"],
    ["sec80C", "sec80CCD1B", "sec80D_self", "sec80D_parents",
     "homeLoanInterest", "hra_exempt", "lta", "sec80TTA_TTB", "sec80G"],
  ];

  // ── Draft & navigation ───────────────────────────────────────
  const next = async () => {
    try {
      await form.validateFields(STEP_FIELDS[current]);
      const values = form.getFieldsValue();
      updateFiling(`step${current}`, values);

      // Serialize dayjs dateOfBirth before JSON transport to avoid circular-ref warning
      const draftValues = { ...values };
      if (draftValues.dateOfBirth?.format) {
        draftValues.dateOfBirth = draftValues.dateOfBirth.format("YYYY-MM-DD");
      }

      await saveDraft({
        itrType:        "ITR-1",
        assessmentYear: "2026-27",
        step:           current,
        data:           { ...filingData, [`step${current}`]: draftValues },
      }).catch(() => {});

      if (current === 2) await computeTaxSummary(values);
      setCurrent((c) => c + 1);
    } catch (_) {}
  };

  const prev = () => setCurrent((c) => c - 1);

  const handleSubmit = async () => {
    if (!taxResult) return;
    setSubmitLoading(true);
    try {
      const s0 = filingData.step0 || {};
      const s1 = filingData.step1 || {};
      const s2 = filingData.step2 || {};

      const payload = {
        selectedRegime: taxResult.betterRegime === "equal" ? "new" : taxResult.betterRegime,
        personalInfo: {
          fullName:          s0.fullName,
          pan:               s0.pan || user?.pan,
          dateOfBirth:       s0.dateOfBirth?.format?.("YYYY-MM-DD") || null,
          gender:            s0.gender,
          residentialStatus: s0.residentialStatus || "ROR",
          city:              s0.city,
          employerName:      s0.employerName,
          employerTAN:       s0.employerTAN,
          bankAccountNo:     s0.bankAccountNo,
          ifscCode:          s0.ifscCode,
        },
        incomeDetails: {
          basicSalary:      s1.basicSalary      || 0,
          hra_received:     s1.hra_received     || 0,
          specialAllowance: s1.specialAllowance || 0,
          bonus:            s1.bonus            || 0,
          professionalTax:  s1.professionalTax  || 0,
          tdsDeducted:      s1.tdsDeducted      || 0,
          interestIncome:   s1.interestIncome   || 0,
          otherIncome:      s1.otherIncome      || 0,
        },
        deductions: {
          sec80C:           s2.sec80C           || 0,
          sec80CCD1B:       s2.sec80CCD1B       || 0,
          sec80D_self:      s2.sec80D_self      || 0,
          sec80D_parents:   s2.sec80D_parents   || 0,
          homeLoanInterest: s2.homeLoanInterest  || 0,
          hra_exempt:       s2.hra_exempt        || 0,
          lta:              s2.lta              || 0,
          sec80TTA_TTB:     s2.sec80TTA_TTB     || 0,
          sec80G:           s2.sec80G           || 0,
        },
      };

      const res = await submitITR1(payload);
      setSubmitted(res.data);
    } catch (err) {
      setError(err.message);
    } finally {
      setSubmitLoading(false);
    }
  };

  const computeTaxSummary = async (values) => {
    setLoading(true);
    setError(null);
    try {
      const allData = { ...filingData.step0, ...filingData.step1, ...values };
      const grossIncome =
        (allData.basicSalary      || 0) +
        (allData.hra_received     || 0) +
        (allData.specialAllowance || 0) +
        (allData.bonus            || 0);
      const payload = {
        grossIncome,
        otherIncome: (allData.interestIncome || 0) + (allData.otherIncome || 0),
        dateOfBirth: allData.dateOfBirth?.format?.("YYYY-MM-DD") || user?.dateOfBirth,
        deductions: {
          sec80C:           allData.sec80C           || 0,
          sec80CCD1B:       allData.sec80CCD1B       || 0,
          sec80D_self:      allData.sec80D_self       || 0,
          sec80D_parents:   allData.sec80D_parents   || 0,
          homeLoanInterest: allData.homeLoanInterest  || 0,
          hra:              allData.hra_exempt        || 0,
          lta:              allData.lta              || 0,
          sec80TTA_TTB:     allData.sec80TTA_TTB     || 0,
          sec80G:           allData.sec80G           || 0,
        },
      };
      const res = await compareRegimes(payload);
      setTaxResult(res.data);
    } catch (err) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  };

  // ── Step 0: Personal Info ────────────────────────────────────
  const PersonalInfo = () => (
    <>
      <Alert
        message="ITR-1 (Sahaj) is for salaried individuals with income up to ₹50 lakhs."
        type="info" showIcon style={{ marginBottom: 24, borderRadius: 8 }}
      />
      <Row gutter={16}>
        <Col xs={24} sm={12}>
          <Form.Item name="fullName" label="Full Name (as per PAN)"
            tooltip="Form 16 → top section: 'Name and address of the Employee'. Copy exactly as printed — it must match your PAN card."
            rules={[{ required: true }]}
            initialValue={user?.fullName}
          >
            <Input prefix={<UserOutlined />} placeholder="RAJESH KUMAR" />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="pan" label="PAN Number"
            tooltip="Form 16 → top section: 'PAN of the Employee / Specified senior citizen'. Pre-filled from your account — verify it matches."
            rules={[{ required: true }]}
            initialValue={user?.pan}
          >
            <Input placeholder="ABCDE1234F" maxLength={10} disabled />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="dateOfBirth" label="Date of Birth"
            tooltip="Not on Form 16 — enter your actual date of birth as per PAN card. Required to determine senior citizen tax slabs."
            rules={[{ required: true, message: "DOB is required" }]}
          >
            <DatePicker style={{ width: "100%" }} format="DD/MM/YYYY" />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="gender" label="Gender"
            tooltip="Not on Form 16 — select as per your PAN card records."
            rules={[{ required: true }]}
          >
            <Select placeholder="Select gender">
              <Option value="M">Male</Option>
              <Option value="F">Female</Option>
              <Option value="T">Transgender</Option>
            </Select>
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="residentialStatus" label="Residential Status"
            tooltip="For most salaried employees in India this is 'Resident & Ordinarily Resident (ROR)'. Select RNOR or NR only if you lived abroad for a significant period during FY 2025-26."
            rules={[{ required: true }]} initialValue="ROR"
          >
            <Select>
              <Option value="ROR">Resident &amp; Ordinarily Resident</Option>
              <Option value="RNOR">Resident but Not Ordinarily Resident</Option>
              <Option value="NR">Non-Resident</Option>
            </Select>
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="city" label="City of Employment"
            tooltip="Your primary city of work during FY 2025-26. Metro cities (Mumbai, Delhi, Kolkata, Chennai) get a higher HRA exemption limit of 50% of basic salary."
            rules={[{ required: true }]}
          >
            <Select placeholder="Select city" showSearch>
              {[...METRO_CITIES, "Bengaluru", "Hyderabad", "Pune", "Ahmedabad", "Other"]
                .map((c) => (
                  <Option key={c} value={c}>
                    {c} {METRO_CITIES.includes(c) ? "🏙 Metro" : ""}
                  </Option>
                ))}
            </Select>
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="fatherName" label="Father's Name"
            tooltip="As per PAN records. Required for ITR verification declaration."
          >
            <Input prefix={<UserOutlined />} placeholder="FATHER'S FULL NAME" />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="aadhaar" label="Aadhaar Number"
            tooltip="12-digit Aadhaar linked to your PAN. Required for EVC (one-click e-verification). Not on Form 16 — check your Aadhaar card."
            rules={[
              { pattern: /^\d{12}$/, message: "Aadhaar must be 12 digits" },
              {
                validator: (_, value) =>
                  !value || isValidAadhaarChecksum(value)
                    ? Promise.resolve()
                    : Promise.reject(new Error("Invalid Aadhaar number — please re-check the digits")),
              },
            ]}
          >
            <Input placeholder="1234 5678 9012" maxLength={12} />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="mobile" label="Mobile (Aadhaar-linked)"
            tooltip="Your mobile number registered with Aadhaar. ITD sends EVC OTP to this number. Not on Form 16."
            rules={[{ pattern: /^[6-9]\d{9}$/, message: "Enter valid 10-digit mobile" }]}
          >
            <Input addonBefore="+91" placeholder="9876543210" maxLength={10} />
          </Form.Item>
        </Col>
        <Col xs={24}>
          <Form.Item name="addressLine1" label="Address (Street / Flat / Colony)"
            tooltip="Your current residential address. Required for ITR filing."
          >
            <Input placeholder="19-4-438/A/10, Street 3, BNK Colony" />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="pinCode" label="PIN Code"
            tooltip="6-digit postal PIN code of your residential area."
            rules={[{ pattern: /^\d{6}$/, message: "PIN must be 6 digits" }]}
          >
            <Input placeholder="500064" maxLength={6} />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12} />
        <Col xs={24}>
          <Form.Item name="employerName" label="Employer Name"
            tooltip="Form 16 → top section: 'Name and address of the Employer / Specified Bank'. Copy the company name exactly."
            rules={[{ required: true }]}
          >
            <Input placeholder="ABC Pvt Ltd" />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="employerTAN" label="Employer TAN"
            tooltip="Form 16 → top section: 'TAN of the Deductor'. A 10-character code like DELM17484F. Required to credit your TDS against your tax liability."
            rules={[
              { required: true },
              { pattern: /^[A-Z]{4}[0-9]{5}[A-Z]{1}$/, message: "Invalid TAN" },
            ]}
          >
            <Input placeholder="ABCD12345E" maxLength={10}
              onChange={(e) => form.setFieldValue("employerTAN", e.target.value.toUpperCase())}
            />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="bankAccountNo" label="Bank Account Number"
            tooltip="Not on Form 16 — this is your bank account where any tax refund will be credited. Use the account linked to your PAN."
            rules={[{ required: true }]}
          >
            <Input placeholder="For refund credit" />
          </Form.Item>
        </Col>
        <Col xs={24} sm={12}>
          <Form.Item name="ifscCode" label="IFSC Code"
            tooltip="Not on Form 16 — the IFSC code of your bank branch. Find it on your cheque book, passbook, or net banking portal."
            rules={[
              { required: true },
              { pattern: /^[A-Z]{4}0[A-Z0-9]{6}$/, message: "Invalid IFSC" },
            ]}
          >
            <Input placeholder="SBIN0001234" maxLength={11}
              onChange={(e) => form.setFieldValue("ifscCode", e.target.value.toUpperCase())}
            />
          </Form.Item>
        </Col>
      </Row>
    </>
  );

  // ── Step 1: Income ───────────────────────────────────────────
  // Salary breakdown matches Form 16 Part B — gross salary is computed from
  // these components rather than entered directly, so it always stays consistent.
  const SALARY_FIELDS = [
    {
      name:     "basicSalary",
      label:    "Basic Salary (₹)",
      required: true,
      tip:      "Form 16 → Annexure (last pages): 'Basic Salary'. Your base pay before any allowances.",
    },
    {
      name:     "hra_received",
      label:    "HRA Received (₹)",
      required: false,
      tip:      "Form 16 → Annexure (last pages): look for 'House Rent Allowance'. This is the total HRA your employer paid you during the year — needed to compute your HRA exemption in the next step. Enter 0 if your employer does not pay HRA.",
    },
    {
      name:     "specialAllowance",
      label:    "Special Allowance (₹)",
      required: false,
      tip:      "Form 16 → Annexure: 'Special Allowance' or 'Other Allowance'. Any fixed allowance beyond Basic and HRA. Enter 0 if not applicable.",
    },
    {
      name:     "bonus",
      label:    "Bonus (₹)",
      required: false,
      tip:      "Form 16 → Annexure: 'Bonus' or 'Performance Pay'. Any bonus or incentive paid during the year. Enter 0 if not applicable.",
    },
  ];

  const OTHER_SALARY_FIELDS = [
    {
      name:     "professionalTax",
      label:    "Professional Tax u/s 16(iii) (₹)",
      required: false,
      tip:      "Form 16 → Part B: 'Tax on employment u/s 16(iii)'. This is the professional tax (Profession Tax) deducted from your salary by your employer and paid to the state government. Usually ₹200/month = ₹2,400/year. Enter 0 if not applicable.",
    },
    {
      name:     "tdsDeducted",
      label:    "TDS Already Deducted (₹)",
      required: true,
      tip:      "Form 16 → Part A summary table: the total 'Amount of tax deposited / remitted' column, OR Part B row labeled 'Tax deducted at source u/s 192(1)'. This is the total TDS your employer deducted across all four quarters.",
    },
  ];

  const OTHER_INCOME_FIELDS = [
    {
      name:  "interestIncome",
      label: "Interest Income (FD/Savings) (₹)",
      tip:   "Not on Form 16 — check your bank's Form 26AS or Annual Information Statement (AIS) for interest credited on FDs and savings accounts during FY 2025-26.",
    },
    {
      name:  "otherIncome",
      label: "Any Other Income (₹)",
      tip:   "Not on Form 16 — include freelance income, rental income, capital gains, or any other taxable receipts not covered above.",
    },
  ];

  const IncomeDetails = () => (
    <>
      <Title level={5}>Salary Income</Title>
      <Row gutter={16}>
        {SALARY_FIELDS.map(({ name, label, required, tip }) => (
          <Col xs={24} sm={12} key={name}>
            <Form.Item name={name} label={label} tooltip={tip}
              rules={required ? [{ required: true, message: `${label} is required` }] : []}
            >
              <InputNumber
                style={{ width: "100%" }} min={0}
                formatter={(v) => `₹ ${v}`.replace(/\B(?=(\d{3})+(?!\d))/g, ",")}
                parser={(v) => v.replace(/₹\s?|(,*)/g, "")}
                placeholder="0"
              />
            </Form.Item>
          </Col>
        ))}
      </Row>

      <Form.Item shouldUpdate noStyle>
        {() => {
          const v = form.getFieldsValue(["basicSalary", "hra_received", "specialAllowance", "bonus"]);
          const total = (v.basicSalary || 0) + (v.hra_received || 0) + (v.specialAllowance || 0) + (v.bonus || 0);
          return (
            <Alert
              type="success"
              showIcon
              message={<Text>Total Gross Salary: <Text strong>{fmt(total)}</Text></Text>}
              style={{ marginBottom: 20, borderRadius: 8 }}
            />
          );
        }}
      </Form.Item>

      <Row gutter={16}>
        {OTHER_SALARY_FIELDS.map(({ name, label, required, tip }) => (
          <Col xs={24} sm={12} key={name}>
            <Form.Item name={name} label={label} tooltip={tip}
              rules={required ? [{ required: true, message: `${label} is required` }] : []}
            >
              <InputNumber
                style={{ width: "100%" }} min={0}
                formatter={(v) => `₹ ${v}`.replace(/\B(?=(\d{3})+(?!\d))/g, ",")}
                parser={(v) => v.replace(/₹\s?|(,*)/g, "")}
                placeholder="0"
              />
            </Form.Item>
          </Col>
        ))}
      </Row>

      <Divider />
      <Title level={5}>Other Income</Title>
      <Row gutter={16}>
        {OTHER_INCOME_FIELDS.map(({ name, label, tip }) => (
          <Col xs={24} sm={12} key={name}>
            <Form.Item name={name} label={label} tooltip={tip}>
              <InputNumber
                style={{ width: "100%" }} min={0}
                formatter={(v) => `₹ ${v}`.replace(/\B(?=(\d{3})+(?!\d))/g, ",")}
                parser={(v) => v.replace(/₹\s?|(,*)/g, "")}
                placeholder="0"
              />
            </Form.Item>
          </Col>
        ))}
      </Row>
    </>
  );

  // ── Step 2: Deductions ───────────────────────────────────────
  const DEDUCTION_FIELDS = [
    {
      name:  "sec80C",
      label: "80C — PF, PPF, ELSS, LIC",
      max:   DEDUCTION_LIMITS.SEC_80C,
      tip:   "Form 16 → Part B: 'Total deduction under section 80C, 80CCC and 80CCD(1)'. Includes Employee PF (shown in Annexure), PPF, ELSS mutual funds, LIC premiums, and home loan principal. Capped at ₹1,50,000.",
    },
    {
      name:  "sec80CCD1B",
      label: "80CCD(1B) — NPS",
      max:   DEDUCTION_LIMITS.SEC_80CCD_1B,
      tip:   "Form 16 → Part B: look for '80CCD(1B)' — additional NPS contribution over and above the 80C limit. Max ₹50,000 extra deduction. Leave blank if you did not contribute to NPS separately.",
    },
    {
      name:  "sec80D_self",
      label: "80D — Health Insurance (Self)",
      max:   DEDUCTION_LIMITS.SEC_80D_SELF,
      tip:   "Form 16 → Part B: 'Deduction in respect of health insurance premia under section 80D'. Covers premium paid for self, spouse, and children. Max ₹25,000 (₹50,000 if you are a senior citizen).",
    },
    {
      name:  "sec80D_parents",
      label: "80D — Health Insurance (Parents)",
      max:   DEDUCTION_LIMITS.SEC_80D_PARENTS,
      tip:   "Not always on Form 16 — enter the health insurance premium you paid for your parents. Max ₹25,000 (₹50,000 if parents are senior citizens). Keep the premium receipt.",
    },
    {
      name:  "homeLoanInterest",
      label: "24(b) — Home Loan Interest",
      max:   200000,
      tip:   "Form 16 → Part B: look for 'Loss from house property' or '24(b)'. Enter the interest portion of your home loan EMIs paid during FY 2025-26. Max ₹2,00,000 for self-occupied property. Get the certificate from your bank.",
    },
    {
      name:  "hra_exempt",
      label: "HRA Exemption",
      max:   null,
      tip:   "Form 16 → Part B: 'House rent allowance under section 10(13A)'. This is the tax-exempt portion of HRA — calculated as the least of: actual HRA received, rent paid minus 10% of basic, or 50%/40% of basic (metro/non-metro). Enter 0 if you live in your own house.",
    },
    {
      name:  "lta",
      label: "LTA Exemption",
      max:   null,
      tip:   "Form 16 → Part B: 'Leave travel concession or assistance under section 10(5)'. Enter the LTA amount your employer has exempted. Covers actual travel fare for domestic trips twice in a 4-year block. Leave blank if not claimed.",
    },
    {
      name:  "sec80TTA_TTB",
      label: "80TTA/TTB — Savings Interest",
      max:   DEDUCTION_LIMITS.SEC_80TTA,
      tip:   "Usually not on Form 16 — check your bank statement or Form 26AS. Enter interest earned on savings accounts (80TTA, max ₹10,000) or all interest if you are a senior citizen (80TTB, max ₹50,000).",
    },
    {
      name:  "sec80G",
      label: "80G — Donations",
      max:   null,
      tip:   "Usually not on Form 16 — enter donations made to eligible charitable organisations (PM Relief Fund, temples, NGOs with 80G registration). Keep the donation receipt. Only the deductible portion (50% or 100% depending on the organisation) applies.",
    },
  ];

  const DeductionsForm = () => (
    <>
      <Alert
        message="Deductions below apply only if you choose the Old Tax Regime. The calculator will show which regime saves more."
        type="warning" showIcon style={{ marginBottom: 20, borderRadius: 8 }}
      />
      <Row gutter={16}>
        {DEDUCTION_FIELDS.map(({ name, label, max, tip }) => (
          <Col xs={24} sm={12} key={name}>
            <Form.Item
              name={name}
              tooltip={tip}
              label={
                <span>
                  {label}
                  {max && (
                    <Text type="secondary" style={{ fontSize: 11, marginLeft: 6 }}>
                      max {fmt(max)}
                    </Text>
                  )}
                </span>
              }
            >
              <InputNumber
                style={{ width: "100%" }} min={0} max={max || undefined}
                formatter={(v) => `₹ ${v}`.replace(/\B(?=(\d{3})+(?!\d))/g, ",")}
                parser={(v) => v.replace(/₹\s?|(,*)/g, "")}
                placeholder="0"
              />
            </Form.Item>
          </Col>
        ))}
      </Row>
    </>
  );

  // ── Step 3: Tax Summary ──────────────────────────────────────
  const TaxSummary = () => {
    if (loading) return <div className="text-center py-10"><Text>Computing your taxes...</Text></div>;
    if (error)   return <Alert type="error" message={error} showIcon />;
    if (!taxResult) return null;

    const { old: o, new: n, betterRegime, savingsAmount } = taxResult;
    const recommended = betterRegime === "new" ? n : o;

    return (
      <>
        <Alert
          message={
            <Text>
              <Text strong>
                {betterRegime === "new" ? "New Regime" : "Old Regime"}
              </Text>{" "}
              saves you{" "}
              <Text strong style={{ color: "#52c41a" }}>{fmt(savingsAmount)}</Text>{" "}
              more in taxes.
            </Text>
          }
          type="success"
          showIcon
          style={{ marginBottom: 20, borderRadius: 8 }}
        />

        <Row gutter={[12, 12]} style={{ marginBottom: 20 }}>
          {[
            { label: "Old Regime", data: o },
            { label: "New Regime", data: n },
          ].map(({ label, data }) => (
            <Col span={12} key={label}>
              <Card
                variant="borderless"
                size="small"
                style={{
                  borderRadius: 10,
                  outline: data === recommended ? `2px solid ${token.colorPrimary}` : `1px solid ${token.colorBorderSecondary}`,
                }}
                title={
                  <Space>
                    <Text strong>{label}</Text>
                    {data === recommended && <Tag color="blue">Recommended</Tag>}
                  </Space>
                }
              >
                {[
                  ["Gross Income",   data.grossIncome],
                  ["Deductions",     data.deductionTotal],
                  ["Taxable Income", data.taxableIncome],
                  ["Rebate 87A",     data.rebateApplied],
                  ["Surcharge",      data.surcharge],
                  ["Cess (4%)",      data.cess],
                  ["Total Tax",      data.totalTax],
                ].map(([lbl, val]) => (
                  <div key={lbl}
                    className="flex justify-between py-1"
                    style={{ borderBottom: `1px solid ${token.colorBorderSecondary}` }}
                  >
                    <Text type="secondary" style={{ fontSize: 12 }}>{lbl}</Text>
                    <Text strong style={{ fontSize: 12,
                      color: lbl === "Total Tax" ? token.colorPrimary : "inherit" }}
                    >
                      {typeof val === "number" ? fmt(val) : val}
                    </Text>
                  </div>
                ))}
                <div className="flex justify-between pt-2">
                  <Text type="secondary" style={{ fontSize: 12 }}>Effective Rate</Text>
                  <Text strong style={{ color: "#52c41a" }}>{data.effectiveRate}%</Text>
                </div>
              </Card>
            </Col>
          ))}
        </Row>

        <Alert
          message="Your ITR-1 draft is ready. Review details and submit to complete e-Filing."
          type="info" showIcon style={{ borderRadius: 8 }}
        />
      </>
    );
  };

  const stepContent = [
    <PersonalInfo   key="0" />,
    <IncomeDetails  key="1" />,
    <DeductionsForm key="2" />,
    <TaxSummary     key="3" />,
  ];

  // ── Submission success screen ────────────────────────────────
  const handleDownloadXML = async () => {
    setXmlLoading(true);
    setXmlError(null);
    try {
      const blob = await downloadFilingXML(submitted.filing._id);
      const url  = URL.createObjectURL(new Blob([blob], { type: "application/xml" }));
      const a   = document.createElement("a");
      a.href     = url;
      a.download = `ITR1_AY2026-27_${submitted.acknowledgementNo}.xml`;
      a.click();
      URL.revokeObjectURL(url);
    } catch {
      setXmlError("Could not generate XML. Please try again.");
    } finally {
      setXmlLoading(false);
    }
  };

  if (submitted) {
    return (
      <div>
        <Result
          status="success"
          icon={<CheckCircleOutlined style={{ color: "#52c41a" }} />}
          title="ITR-1 Prepared Successfully!"
          subTitle={
            <div>
              <p>App Acknowledgement: <Text code strong>{submitted.acknowledgementNo}</Text></p>
              <p>Assessment Year: <strong>AY 2026-27</strong></p>
              <p>
                Total Tax:{" "}
                <strong style={{ color: "#fa541c" }}>{fmt(submitted.taxSummary?.totalTax)}</strong>
                {submitted.taxSummary?.refundDue > 0 && (
                  <span style={{ marginLeft: 12, color: "#52c41a" }}>
                    · Refund: {fmt(submitted.taxSummary?.refundDue)}
                  </span>
                )}
              </p>
            </div>
          }
          extra={[
            <Button
              type="primary"
              key="xml"
              icon={<DownloadOutlined />}
              loading={xmlLoading}
              onClick={handleDownloadXML}
            >
              Download ITR XML
            </Button>,
            <Button
              key="efiling"
              icon={<SafetyCertificateOutlined />}
              onClick={() => navigate("/efiling")}
            >
              e-File via Platform
            </Button>,
            <Button key="dashboard" onClick={() => navigate("/dashboard")}>
              Go to Dashboard
            </Button>,
          ]}
        />

        {xmlError && (
          <Alert type="error" message={xmlError} showIcon style={{ marginBottom: 16, borderRadius: 8 }} />
        )}

        {/* Step-by-step guide for manual upload */}
        <Card
          style={{ maxWidth: 680, margin: "0 auto", borderRadius: 10 }}
          variant="borderless"
        >
          <Alert
            type="info"
            showIcon
            icon={<GlobalOutlined />}
            message={<strong>How to file on the Income Tax Portal (incometax.gov.in)</strong>}
            description={
              <ol style={{ paddingLeft: 20, margin: "8px 0 0", lineHeight: 2 }}>
                <li>Click <strong>Download ITR XML</strong> above to save your pre-filled return</li>
                <li>
                  Go to{" "}
                  <a href="https://www.incometax.gov.in/iec/foportal/" target="_blank" rel="noreferrer">
                    incometax.gov.in
                  </a>{" "}
                  and log in with your PAN and password
                </li>
                <li>Navigate to <strong>e-File → Income Tax Returns → File Income Tax Return</strong></li>
                <li>Select <strong>AY 2026-27</strong>, then <strong>ITR-1</strong>, then <strong>Upload ITR</strong></li>
                <li>Choose <strong>Upload XML</strong> and select the downloaded file</li>
                <li>Verify using <strong>Aadhaar OTP</strong> or <strong>Net Banking EVC</strong> to complete filing</li>
                <li>Save the <strong>ITR-V Acknowledgement</strong> sent to your registered email</li>
              </ol>
            }
            style={{ borderRadius: 8 }}
          />

          <Divider />

          <div style={{ display: "flex", justifyContent: "space-between", alignItems: "center", flexWrap: "wrap", gap: 8 }}>
            <Text type="secondary" style={{ fontSize: 12 }}>
              Or use our integrated e-filing to submit directly without leaving the app
            </Text>
            <Button
              type="link"
              icon={<SafetyCertificateOutlined />}
              onClick={() => navigate("/efiling")}
              style={{ padding: 0 }}
            >
              e-File via Platform →
            </Button>
          </div>
        </Card>
      </div>
    );
  }

  // Drawer width: full-screen on mobile, 680px on desktop
  const drawerWidth = screens.lg ? 680 : "100%";

  return (
    <div>
      <PageHeader
        icon={<FileTextOutlined />}
        title="ITR-1 Filing — Sahaj"
        subtitle="Salaried individuals | FY 2025-26 | AY 2026-27"
        extra={
          <Tooltip
            title={
              docsLoading  ? "Checking for Form 16…" :
              form16Doc    ? `Open ${form16Doc.originalName} for reference` :
                             "No Form 16 uploaded yet — upload one below"
            }
          >
            <Button
              icon={docsLoading ? <Spin size="small" /> : <PaperClipOutlined />}
              onClick={() => setDrawerOpen(true)}
              disabled={docsLoading}
              style={{
                borderColor: form16Doc ? "#1677ff" : undefined,
                color:       form16Doc ? "#1677ff" : undefined,
              }}
            >
              {screens.sm ? "View Form 16" : ""}
            </Button>
          </Tooltip>
        }
      />

      {/* Steps indicator */}
      <Card variant="borderless" style={{ borderRadius: 10, marginBottom: 24 }}>
        <Steps current={current} items={STEPS} />
      </Card>

      {/* Step content */}
      <Card variant="borderless" style={{ borderRadius: 10, marginBottom: 16 }}>
        <Form form={form} layout="vertical" size="large">
          {stepContent[current]}
        </Form>
      </Card>

      {/* Navigation */}
      <Card variant="borderless" style={{ borderRadius: 10 }}>
        <Row justify="space-between">
          <Col>
            {current > 0 && (
              <Button icon={<ArrowLeftOutlined />} onClick={prev} size="large">
                Previous
              </Button>
            )}
          </Col>
          <Col>
            {current < STEPS.length - 1 ? (
              <Button
                type="primary"
                icon={<ArrowRightOutlined />}
                onClick={next}
                size="large"
                loading={loading && current === 2}
              >
                {current === 2 ? "Compute Tax" : "Next"}
              </Button>
            ) : (
              <Button
                type="primary"
                icon={<CheckCircleOutlined />}
                size="large"
                loading={submitLoading}
                onClick={handleSubmit}
                disabled={!taxResult}
              >
                Submit ITR-1
              </Button>
            )}
          </Col>
        </Row>
      </Card>

      {/* ── Form 16 Reference Drawer ─────────────────────────── */}
      <Drawer
        open={drawerOpen}
        onClose={() => setDrawerOpen(false)}
        width={drawerWidth}
        title={
          <Space>
            <FilePdfOutlined style={{ color: "#ff4d4f" }} />
            <span>
              {form16Doc ? form16Doc.originalName : "Form 16 Reference"}
            </span>
            {form16Doc && (
              <Tag color="blue" style={{ fontSize: 11 }}>
                FY {form16Doc.financialYear}
              </Tag>
            )}
          </Space>
        }
        styles={{ body: { padding: 0, display: "flex", flexDirection: "column" } }}
      >
        {form16Doc ? (
          <>
            <div style={{
              padding: "12px 16px", borderBottom: `1px solid ${token.colorBorderSecondary}`,
              display: "flex", justifyContent: "space-between", alignItems: "center",
              flexWrap: "wrap", gap: 8,
            }}>
              <Text type="secondary" style={{ fontSize: 12 }}>
                {form16Doc.parsedData
                  ? `Parsed with ${form16Doc.parsedData.confidence ?? 0}% confidence`
                  : "Uploaded"}
              </Text>
              <Space>
                {form16Doc.parsedData && (
                  <Button type="primary" size="small" icon={<ThunderboltOutlined />} onClick={handleUseForm16Values}>
                    Use These Values
                  </Button>
                )}
                <Upload
                  accept=".pdf"
                  multiple={false}
                  showUploadList={false}
                  disabled={uploading}
                  beforeUpload={(file) => { handleForm16Upload({ file }); return false; }}
                >
                  <Button size="small" icon={<UploadOutlined />} loading={uploading}>Replace</Button>
                </Upload>
              </Space>
            </div>
            <iframe
              src={`/uploads/${form16Doc.storedName}`}
              title="Form 16 PDF"
              style={{ flex: 1, width: "100%", height: "100%", border: "none", minHeight: "70vh" }}
            />
          </>
        ) : (
          <div style={{ padding: 32 }}>
            {uploading ? (
              <div style={{ textAlign: "center", padding: 40 }}>
                <Spin size="large" />
                <p style={{ marginTop: 12 }}>Uploading and parsing Form 16...</p>
              </div>
            ) : (
              <Dragger
                accept=".pdf"
                multiple={false}
                showUploadList={false}
                beforeUpload={(file) => { handleForm16Upload({ file }); return false; }}
              >
                <p className="ant-upload-drag-icon">
                  <InboxOutlined style={{ color: "#1677ff" }} />
                </p>
                <p className="ant-upload-text">Click or drag your Form 16 PDF here</p>
                <p className="ant-upload-hint">PDF only · Max 5MB · FY 2025-26 — we'll parse it and offer to fill the form for you</p>
              </Dragger>
            )}
          </div>
        )}
      </Drawer>
    </div>
  );
}
